Move CMSDK Counters/Timers to unified interface

fvincenzo opened the pull request:
The CMSDK Timers (Single/Dual) can be used as a timer or as a counter. The unified interface proposed in #8340 unifies counter.h and rtc.h to provide a common interface.

This patchset makes the CMSDK Timers drivers compliant with the new proposed interface.

**Changes:**
**v5:**
 - Addressed review comments.

**v4:**
 - Implemented the new set_wrap/set_alarm API
 - Fixed a bug
 - Rebase on the latest **topic-counters** branch
 - General code cleanup

**v3:**
 - Removed status check in set_wrap_alarm
 - Rebase on latest master

**v2:**
 - Updated device tree for mps2_an385. The timers and counters are now working correctly even of this platform.

nashif commented:
good to see you back @fvincenzo :)

nordic-krch commented on drivers/counter/timer_dtmr_cmsdk_apb.c:
```diff
@@ -87,9 +87,10 @@ static u32_t timer_dtmr_cmsdk_apb_read(struct device *dev)
+static int timer_dtmr_cmsdk_apb_set_wrap_alarm(struct device *dev,
+					       u32_t ticks,
+					       counter_wrap_callback_t callback,
+					       void *user_data)
```
note that there is a change in API allowing to set wrap even when counter is not started yet (https://github.com/zephyrproject-rtos/zephyr/blob/6d0a822ebd54059b154458c63bae6216326a344a/include/counter.h#L317). So `if` from line 99 could be removed.

fvincenzo replied:
Agreed. I will address this in v3

nordic-krch commented on drivers/counter/timer_tmr_cmsdk_apb.c:
```diff
@@ -87,9 +87,10 @@ static u32_t timer_tmr_cmsdk_apb_read(struct device *dev)
+static int timer_tmr_cmsdk_apb_set_wrap_alarm(struct device *dev,
+					      u32_t ticks,
+					      counter_wrap_callback_t callback,
+					      void *user_data)
```
same here. `if (data->status == TIMER_DISABLED) {` can be removed.

fvincenzo replied:
Agreed. I will address this in v3

nordic-krch commented:
please fix conflicts

nordic-krch commented:
please rebase on topic-counters

galak commented:
> please rebase on topic-counters

rebased.

fvincenzo commented:
> > please rebase on topic-counters
> 
> rebased.

@galak Thanks :+1:

fvincenzo commented:
> please rebase on topic-counters

@nordic-krch  I did the changes your required in the last iterations of these patches, could you please acknowledge you change requests and close them?

nordic-krch commented on drivers/counter/timer_dtmr_cmsdk_apb.c:
```diff
@@ -87,26 +88,61 @@ static u32_t timer_dtmr_cmsdk_apb_read(struct device *dev)
 
 	/* Store the reload value */
-	data->load = count;
+	data->load = alarm_cfg->ticks;
```
that looks suspicious. load value looks like wrap (and it is modified in `set_wrap` function). `set_alarm` should configure compare channels if timer peripheral has any. Note that `set_wrap` is equivalent of legacy `set_alarm` function and `set_channel_alarm` is something new.

nordic-krch commented on drivers/counter/timer_dtmr_cmsdk_apb.c:
```diff
@@ -128,21 +176,28 @@ static const struct counter_driver_api timer_dtmr_cmsdk_apb_api = {
+		(*(data->user_cb))(dev, NULL);
 	}
 
+	if (data->user_handler) {
```
seems to me like on wrap is supperte since there is no compare channels. That should be removed.

nordic-krch commented on drivers/counter/timer_dtmr_cmsdk_apb.c:
```diff
@@ -128,21 +176,28 @@ static const struct counter_driver_api timer_dtmr_cmsdk_apb_api = {
 	cfg->dtimer->timer1intclr = DUALTIMER_INTCLR;
-	if (user_cb) {
+	if (data->user_cb) {
 		/* user_data paramenter is not used by this driver */
```
seems like missing implementation. It's software feature so should be easily supported.

nordic-krch commented on drivers/counter/timer_tmr_cmsdk_apb.c:
```diff
@@ -87,36 +89,86 @@ static u32_t timer_tmr_cmsdk_apb_read(struct device *dev)
+	return ticks;
+}
+
 static int timer_tmr_cmsdk_apb_set_alarm(struct device *dev,
```
similar thing as in other driver, it seems like `set_wrap` should not be implemented. only `set_wrap`

nordic-krch commented on drivers/counter/timer_tmr_cmsdk_apb.c:
```diff
@@ -130,19 +182,27 @@ static const struct counter_driver_api timer_tmr_cmsdk_apb_api = {
+	if (data->user_cb) {
 		/* user_data paramenter is not used by this driver */
-		(*user_cb)(dev, NULL);
+		(*(data->user_cb))(dev, NULL);
```
see comments in another driver.

MaureenHelm commented on drivers/counter/counter_dtmr_cmsdk_apb.c:
```diff
@@ -63,9 +63,28 @@ static u32_t counter_dtmr_cmsdk_apb_read(struct device *dev)
+
+static u32_t counter_dtmr_cmsdk_apb_get_wrap(struct device *dev)
+{
+	return 0;
```
Shouldn't this be the last value before the wrap? e.g., 0xFFFF for a 16-bit counter?

fvincenzo replied:
Good catch! Thanks @MaureenHelm :+1:  Solved in the new version.

MaureenHelm commented on drivers/counter/timer_dtmr_cmsdk_apb.c:
```diff
@@ -87,26 +88,61 @@ static u32_t timer_dtmr_cmsdk_apb_read(struct device *dev)
-		return -ENOTSUP;
-	}
+	if (alarm_cfg->handler == NULL)
+		return -ENODEV;
```
This line needs to be wrapped in braces

fvincenzo replied:
Done. Thanks

MaureenHelm commented:
@fvincenzo 	It looks like some non-counter related stuff slipped into this PR since you submitted it from fvincenzo:master. Will you please move the counter patches to their own branch and resubmit the PR?

fvincenzo commented:
> @fvincenzo It looks like some non-counter related stuff slipped into this PR since you submitted it from fvincenzo:master. Will you please move the counter patches to their own branch and resubmit the PR?

@MaureenHelm yes, sure, I can do it.

fvincenzo commented:
> > @fvincenzo It looks like some non-counter related stuff slipped into this PR since you submitted it from fvincenzo:master. Will you please move the counter patches to their own branch and resubmit the PR?
> 
> @MaureenHelm yes, sure, I can do it.

@MaureenHelm Some patches were removed from my original PR and this was rebased on `origin/master` instead of `nordic-krch/topic-counters` as originally meant. The result is that I can't resubmit the PR in a coherent way unless `nordic-krch/topic-counters` is rebased on `origin/master`. @nordic-krch Could you please do that?

MaureenHelm commented:
> @MaureenHelm Some patches were removed from my original PR and this was rebased on `origin/master` instead of `nordic-krch/topic-counters` as originally meant. The result is that I can't resubmit the PR in a coherent way unless `nordic-krch/topic-counters` is rebased on `origin/master`. @nordic-krch Could you please do that?

Why can't you rebase onto zephyrproject-rtos/topic-counters? The topic branch was rebased onto master earlier this week.

fvincenzo commented:
> > @MaureenHelm Some patches were removed from my original PR and this was rebased on `origin/master` instead of `nordic-krch/topic-counters` as originally meant. The result is that I can't resubmit the PR in a coherent way unless `nordic-krch/topic-counters` is rebased on `origin/master`. @nordic-krch Could you please do that?
> 
> Why can't you rebase onto zephyrproject-rtos/topic-counters? The topic branch was rebased onto master earlier this week.
@MaureenHelm there was a patch overlapping, which now seems solved with the last rebase. Thanks @nordic-krch

nordic-krch commented on drivers/counter/counter_dtmr_cmsdk_apb.c:
```diff
@@ -75,6 +94,9 @@ static const struct counter_driver_api counter_dtmr_cmsdk_apb_api = {
 	.read = counter_dtmr_cmsdk_apb_read,
 	.set_alarm = counter_dtmr_cmsdk_apb_set_alarm,
+	.disable_alarm = counter_dtmr_cmsdk_apb_disable_alarm,
+	.set_wrap = counter_dtmr_cmsdk_apb_set_wrap_alarm,
```
please align to latest changes `set_wrap` -> `set_top_value`, `get_wrap`->`get_top_value`

MaureenHelm replied:
The fact that CI is still passing suggests that we're not building this driver in CI. @fvincenzo 	will you please add `counter` to the supported list in the appropriate \<board\>.yaml file? Like this: https://github.com/zephyrproject-rtos/zephyr/blob/817c74fbd0546f10d5ff5f66d5a1aeeb09f8b575/boards/arm/frdm_kw41z/frdm_kw41z.yaml#L9-L12

nordic-krch commented on drivers/counter/counter_dtmr_cmsdk_apb.c:
```diff
@@ -63,9 +63,28 @@ static u32_t counter_dtmr_cmsdk_apb_read(struct device *dev)
+	return DUALTIMER_MAX_RELOAD;
+}
+
 static int counter_dtmr_cmsdk_apb_set_alarm(struct device *dev,
```
if `counter_config_info` is correctly set you don't need to provide this function as it will be never called if counter_config_info indicates that there is not channels. See `counter_set_channel_alarm` in counter.h

nordic-krch commented on drivers/counter/timer_tmr_cmsdk_apb.c:
```diff
@@ -15,7 +15,7 @@
 typedef void (*timer_config_func_t)(struct device *dev);
 
-static counter_callback_t user_cb;
+static counter_wrap_callback_t user_cb;
```
name has changed.

nordic-krch commented:
@fvincenzo i added couple of minor comments. Please rebase on latest topic-counters and update the code. I think that build is green because this driver is never compiled. You could try to add a board to test so that it's verified.

fvincenzo commented:
> @fvincenzo that is indeed strange. counter api struct has no `.set_wrap` field anymore (it's renamed to `set_top_value`):
> [zephyr/include/counter.h](https://github.com/zephyrproject-rtos/zephyr/blob/26b51369ea761adb21064213372a1e7c29580170/include/counter.h#L111)
> 
> Line 111 in [26b5136](/zephyrproject-rtos/zephyr/commit/26b51369ea761adb21064213372a1e7c29580170)
> 
>  counter_api_set_top_value set_top_value; 
> And in your code you are setting it in the driver. So would expect that compilation will fail

@nordic-krch  Well then we have to work out why this is happening :smile:. So are we sure that the topic branch is 'origin/topic-counters'?

nordic-krch commented:
@fvincenzo you are missing 6 commits in origin/topic-counters: https://github.com/zephyrproject-rtos/zephyr/commits/topic-counters

fvincenzo commented:
> @fvincenzo i added couple of minor comments. Please rebase on latest topic-counters and update the code. I think that build is green because this driver is never compiled. You could try to add a board to test so that it's verified.

@nordic-krch This commits were put on the tree after I updated my pull request (this explains)... Anyway, since the interface changed several times from last September (and I have been rebasing few times to accomodate it), how about you freeze it and than we start the porting of the drivers, otherwise it will always be a race behind you... :smile:

nordic-krch commented:
@fvincenzo sorry for that, unfortunately for driver maintainers there are still active discussions so api will change again but we want to get to master with current api and do the second round once there is a conclusion.

fvincenzo commented:
> @fvincenzo sorry for that, unfortunately for driver maintainers there are still active discussions so api will change again but we want to get to master with current api and do the second round once there is a conclusion.

@nordic-krch That is fine by me. Changing an API is a complex process... and it is better to get it right. From the process stand point, I believe that the drivers porting should start only when the API reaches a versioned (i.e. 1.0) state. So that a driver can be declared compliant with a certain version. When are you going to release the first version?

nordic-krch commented:
could you rebase your PR so we can merge it? I see that it contains additional 21 commits that shouldn't be there.

galak commented:
> could you rebase your PR so we can merge it? I see that it contains additional 21 commits that shouldn't be there.

rebased

nordic-krch commented:
> how about you freeze it and than we start the porting of the drivers, otherwise it will always be a race behind you... 😄

@fvincenzo It's freezed for now until we go to master. please apply comments so we can merge it.

galak committed:
tests: counter_basic_api: Enable tests for ARM CMSDK drivers

Enable testing of counter (COUNTER_{D}TMR_CMSDK) on mps2_an385 and
enable timers on v2m_beetle (TIMER_{D}TMR_CMSDK).

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: counter: cmsdk: Convert to new DT_<COMPAT>_<INSTANCE> defines

Convert cmsdk driver to use new defines so we can remove the
dts_fixup.h code for it.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
dts: arm_cmsdk_(d)timer: Add label property to (d)timer

Add a label property so we can use that in drivers rather than getting
from Kconfig.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
@MaureenHelm, @nordic-krch can you guys re-review.

nordic-krch commented on drivers/counter/timer_dtmr_cmsdk_apb.c:
```diff
@@ -115,6 +112,33 @@ static int timer_dtmr_cmsdk_apb_set_alarm(struct device *dev,
+	return ticks;
+}
+
+static int timer_dtmr_cmsdk_apb_set_alarm(struct device *dev, u8_t chan_id,
```
Driver should provide info structure on driver initialization (https://github.com/zephyrproject-rtos/zephyr/blob/4d8f735fc635d51491d71a489c65c3b94d67a209/include/counter.h#L84). Info structure contains information about number of channels. If channels = 0 then `set_channel_alarm` and `disable_channel_alarm` might not be provided.

galak replied:
can you explain further what should be done here?

nordic-krch replied:
you can see for example here: https://github.com/NXPmicro/zephyr/blob/d880c8a83c4a78d1be099635925249eddd20353a/drivers/counter/counter_mcux_rtc.c#L253

that there is a info structure that should be created with driver instance. Then this struct is used by coutner API for geting for example number of channels. `counter_set_channel_alarm` is using this function to check if given instance supports setting set_channel_alarm. If instance has no channels then it does not need to implement `set_channel` and `disable_channel`

galak replied:
Gotcha, how is a channel defined?

nordic-krch replied:
what do you mean?

galak replied:
@nordic-krch Trying to understand what the definition of an channel is.

nordic-krch replied:
channel usually maps to compare channel seen in timer peripheral. It is able to generate interrupt on given tick within counter period (period is being set by `set_top_value`). In pure hardware driver number of channels is usually limited by number of compare channels in the peripheral. In theory one could build multiple channels on single physical compare channel.

galak replied:
And how is an alarm defined?

nordic-krch replied:
currently this function is called in api `set_channel_alarm` so alarm is set for given channel. It's been renamed from `counter_set_alarm` because there is already deprecated function named `counter_set_alarm`.

galak replied:
Is there anything that distinguishes a basic counter from an alarm?

On the CMSDK dual-timer hardware, trying to decide if I should treat the 2nd timer as just another counter or as an alarm.

nordic-krch replied:
i would treat those as two counters. There is a dependency between running counter and alarm. e.g. alarm can be relative to current counter value or absolute. So if you have 2 timers which are independent and they do not have compare channels to be those are 2 counters.

nordic-krch commented on drivers/counter/counter_tmr_cmsdk_apb.c:
```diff
@@ -63,18 +63,45 @@ static u32_t counter_tmr_cmsdk_apb_read(struct device *dev)
+	return -ENODEV;
+}
+
+static int counter_tmr_cmsdk_apb_disable_alarm(struct device *dev, u8_t chan_id)
```
Driver should provide info structure on driver initialization (https://github.com/zephyrproject-rtos/zephyr/blob/4d8f735fc635d51491d71a489c65c3b94d67a209/include/counter.h#L84). Info structure contains information about number of channels. If channels = 0 then `set_channel_alarm` and `disable_channel_alarm` might not be provided.

nordic-krch replied:
note that counter api expects that info structure is set as it calls it unconditionally: https://github.com/zephyrproject-rtos/zephyr/blob/4d8f735fc635d51491d71a489c65c3b94d67a209/include/counter.h#L141

nordic-krch commented on drivers/counter/timer_tmr_cmsdk_apb.c:
```diff
@@ -87,36 +87,60 @@ static u32_t timer_tmr_cmsdk_apb_read(struct device *dev)
+	return ticks;
+}
+
+static int timer_tmr_cmsdk_apb_set_alarm(struct device *dev, u8_t chan_id,
```
same as in other implementations. Function is not needed once `struct counter_config_info` is provided.

galak committed:
drivers: timer: Move Dual Timer API to unified interface

The CMSDK Dual Timer can be used as a timer or as a counter.
The unified interface proposed in #8340 unifies counter.h and rtc.h to
provide a common interface.

This patch modifies the timer implementation of the dual timer to
make it compliant with the new proposed interface.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: timer: Move Timer API to unified interface

The CMSDK Timer can be used as a timer or as a counter.
The unified interface proposed in #8340 unifies counter.h and rtc.h to
provide a common interface.

This patch modifies the timer implementation of the single timer to
make it compliant with the new proposed interface.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
@nordic-krch @MaureenHelm can you guys re-review.

nordic-krch reviewed (approved):
approve with 2 minor comments (seems like description is not updated after taking it from template).

galak commented on dts/bindings/timer/arm,cmsdk-dtimer.yaml:
```diff
@@ -24,4 +24,10 @@ properties:
+    label:
+      type: string
+      category: required
+      description: Human readable string describing the device (used by Zephyr for API name)
```
Meaning?
